components/service/glean-gradle-plugin/src/main/groovy/mozilla/components/service/glean-gradle-plugin/GleanGradlePlugin.groovy

```diff
@@ -0,0 +1,287 @@
+                // We intentionally don't let out the local metrics.yaml, even if another one will
+                // be extracted from the AAR file. The parser will just ignore this one, if not present.
+                args "${project.projectDir}/metrics.yaml"
+                args "${project.projectDir}/pings.yaml"
+
+                // If we're building the Glean library itself (rather than an
+                // application using Glean) pass the --allow-reserved flag so we can
+                // use metrics in the "glean..." category
+                if (project.name == 'service-glean' || project.ext.has("allowGleanInternal")) {
+                    args "--allow-reserved"
+                }
+
+                doFirst {
+                    // Add the potential 'metrics.yaml' files at evaluation-time, rather than
+                    // configuration-time. Otherwise the Gradle build will fail.
+                    if (project.ext.has("allowMetricsFromAAR")) {
+                        inputs.files.forEach { file ->
+                            project.logger.lifecycle("Glean SDK - generating API from ${file.path}")
+                            args file.path
+                        }
+                    }
+                }
+
+                // Only show the output if something went wrong.
+                ignoreExitValue = true
+                standardOutput = new ByteArrayOutputStream()
+                errorOutput = standardOutput
+                doLast {
+                    if (execResult.exitValue != 0) {
+                        throw new GradleException("Process '${commandLine}' finished with non-zero exit value ${execResult.exitValue}:\n\n${standardOutput.toString()}")
+                    }
+                }
+            }
+
+            // Only attach the generation task if the metrics file is available or we're requested
+            // to fetch them from AAR files. We don't need to fail hard otherwise, as some 3rd party
+            // project might just want metrics included in Glean and nothing more.
+            if (project.file("${project.projectDir}/metrics.yaml").exists()
+                || project.file("${project.projectDir}/pings.yaml").exists()
+                || project.ext.has("allowMetricsFromAAR")) {
+                // This is an Android-Gradle plugin 3+-ism.  Culted from reading the source,
+                // searching for "registerJavaGeneratingTask", and finding
+                // https://github.com/GoogleCloudPlatform/endpoints-framework-gradle-plugin/commit/2f2b91476fb1c6647791e2c6fe531a47615a1e85.
+                // The added directory doesn't appear in the paths listed by the
+                // `sourceSets` task, for reasons unknown.
+                variant.registerJavaGeneratingTask(generateKotlinAPI, new File(sourceOutputDir))
+            }
+        }
+    }
+
+    void apply(Project project) {
+        // This installs a Miniconda3 environment into the gradle user home directory
```

Comment:
This feels like the wrong time to do this.  Plugin application is very early.  You want to do this as part of some task, surely -- otherwise I can't run `./gradlew app:tasks` without installing Python, etc.

Reply:
That's a good point.  I'll see if I can move this inside of `generateMetricsAPI` (where it is definitely needed).

Reply:
It looks like this is actually happening at the right time.  This is just setting up tasks (the most important of which is actually defined by JetBrain's Python Gradle plugin) and installs it into the work tree with `project.preBuild.finalizedBy("build_envs")` below.  I think `preBuild` is the right thing to hang off of, since we do need to have the generated code ready *before* build proper.

I tested by removing the Python environment (in `~/.gradle/glean`) and simple Gradle tasks like `dependencies` work without installing the whole Python environment.

So I'd say we just need to clarify this comment...

Reply:
OK, I see that the paths can be computed without actually doing any work.  Roll on.